`Development`: Document Jenkins and LocalVC setup (#9266)
ls1intum/Artemis: docs/dev/setup/jenkins-localvc.rst

Reviewer: ToDo @b-fein: Restructure Jenkins setup documentation

Right now the documentation jumps frequently between steps required only for local setups, then steps required for production setups, some Nginx setup in between. This is *very* hard to understand.

Not part of this PR since for now it is more important to reuse the existing setup docs for the deprecated GitLab+Jenkins setup to for now retain the important information when the GitLab part gets removed.

Author: --> Issue: https://github.com/ls1intum/Artemis/issues/9273

the file ``application-prod.yml`` (Production Server) or
``application-artemis.yml`` (Local Development) accordingly. Please note
that all values in ``<..>`` have to be configured properly. These values
will be explained below in the corresponding sections. If you want to set up a local environment, copy the values
below into your ``application-artemis.yml`` or ``application-local.yml`` file (the latter is recommended).

.. code:: yaml

   artemis:
    course-archives-path: ./exports/courses
    repo-clone-path: ./repos
    repo-download-clone-path: ./repos-download
    bcrypt-salt-rounds: 11  # The number of salt rounds for the bcrypt password hashing. Lower numbers make it faster but more unsecure and vice versa.
                            # Please use the bcrypt benchmark tool to determine the best number of rounds for your system. https://github.com/ls1intum/bcrypt-Benchmark
    user-management:
        use-external: false
        internal-admin:
            username: artemis_admin
            password: artemis_admin
        accept-terms: false
        login:
            account-name: TUM
    version-control:

>         url: http://172.17.0.1:8080 # `http://host.docker.internal:8080` for Windows

Reviewer 2: This is shown to the users as the clone url. Maybe use `localhost` here.

Author: Addressed with 57097a78ed99417f7d813fa3ea22dcd4d20036d2

        user: root
        password: dummy # has to be set, but does not matter for LocalVC
        build-agent-git-username: jenkins
        build-agent-git-password: artemis_admin # choose some strong password and username (gives read access to all repositories)
    continuous-integration:
        user: artemis_admin
        password: artemis_admin
        url: http://localhost:8082
        vcs-credentials: artemis_localvc_credentials
        artemis-authentication-token-key: artemis_notification_plugin_token
        artemis-authentication-token-value: artemis_admin
        build-timeout: 30
    git:
        name: Artemis
        email: artemis@xcit.tum.de
   jenkins:
       # only required if Artemis and Jenkins cannot communicate on their public URLs
       # e.g., Jenkins is only available in a local container network
       internal-urls:
           ci-url: http://jenkins:8080

>            vcs-url: http://localhost:8080

Reviewer 2: See other `vcs-url` comment

Reviewer 3: for internal that should be artemis:8080 or something similar :)

Reviewer 2: That's assuming Artemis runs in a container. The other uses of `172.17.0.1` imply that it would not.

Author: Addressed with a7828f5915b09b735b666c1dd3f584499c9acd35

This is ideal as a quickstart for developers. For a more detailed setup, see
`Manual Jenkins Server Setup <#manual-jenkins-server-setup>`__.
In a production setup, you have to at least change the user credentials (in the file ``jenkins-casc-config-localvc.yml``).

#. You can now first build and deploy Jenkins, then you can also start the other services which weren't started yet:

    .. code:: bash

       JAVA_OPTS=-Djenkins.install.runSetupWizard=false docker compose -f docker/<Jenkins setup to be launched>.yml up --build -d jenkins
       docker compose -f docker/<Jenkins setup to be launched>.yml up -d

   Jenkins is then reachable under ``http://localhost:8082/`` and you can login using the credentials specified
   in ``jenkins-casc-config-localvc.yml`` (defaults to ``artemis_admin`` as both username and password).

#. The `application-local.yml` must be adapted with the values configured in ``jenkins-casc-config-localvc.yml``:

    .. code:: yaml

        artemis:
            user-management:
                use-external: false
                internal-admin:
                    username: artemis_admin
                    password: artemis_admin
            version-control:

>                 url: http://172.17.0.1:8080 # `http://host.docker.internal:8080` for Windows

Reviewer 2: See other `version-control.url` comment

Author: Addressed with 57097a78ed99417f7d813fa3ea22dcd4d20036d2

                user: root
                password: dummy # have to be set, but does not matter for LocalVC
                build-agent-git-username: jenkins
                build-agent-git-password: artemis_admin # choose some strong password and username (gives read access to all repositories)
            continuous-integration:
                user: artemis_admin
                password: artemis_admin
                url: http://localhost:8082
                vcs-credentials: artemis_localvc_credentials
                artemis-authentication-token-key: artemis_notification_plugin_token
                artemis-authentication-token-value: artemis_admin

#. Open the ``src/main/resources/config/application-jenkins.yml`` and change the following:
   Again, if you are using a development setup, the template in the beginning of this page already contains the
   correct values.

    .. code:: yaml

        jenkins:
            internal-urls:
                ci-url: http://jenkins:8080

>                 vcs-url: http://localhost:8080

Reviewer 2: `jenkins.internal-urls.vcs-url` is used as the clone url by the Jenkins container. It doesn't reach LocalVC at its `localhost`. Use `http://172.17.0.1:8080` or `http://host.docker.internal:8080` on Windows.

Author: Addressed with a7828f5915b09b735b666c1dd3f584499c9acd35

``application-artemis.yml`` (Local Development) accordingly. Please note
that all values in ``<..>`` have to be configured properly. These values
will be explained below in the corresponding sections. If you want to set up a local environment, copy the values
below into your ``application-artemis.yml`` or ``application-local.yml`` file (the latter is recommended).

.. code:: yaml

   artemis:
    course-archives-path: ./exports/courses
    repo-clone-path: ./repos
    repo-download-clone-path: ./repos-download
    bcrypt-salt-rounds: 11  # The number of salt rounds for the bcrypt password hashing. Lower numbers make it faster but more unsecure and vice versa.
                            # Please use the bcrypt benchmark tool to determine the best number of rounds for your system. https://github.com/ls1intum/bcrypt-Benchmark
    user-management:
        use-external: false
        internal-admin:
            username: artemis_admin
            password: artemis_admin
        accept-terms: false
        login:
            account-name: TUM
    version-control:
        url: http://172.17.0.1:8080 # `http://host.docker.internal:8080` for Windows
        user: root

>         password: dummy # have to be set, but does not matter for LocalVC

Reviewer: ```suggestion
        password: dummy # has to be set, but does not matter for LocalVC
```

        build-agent-git-username: jenkins
        build-agent-git-password: artemis_admin # choose some strong password and username (gives read access to all repositories)
    continuous-integration:
        user: artemis_admin
        password: artemis_admin
        url: http://localhost:8082

>         secret-push-token: AQAAABAAAAAg/aKNFWpF9m2Ust7VHDKJJJvLkntkaap2Ka3ZBhy5XjRd8s16vZhBz4fxzd4TH8Su # pre-generated or replaced in Automated Jenkins Server step 3

Reviewer: No longer required like in the GitLab setup?

Author: Addressed with https://github.com/ls1intum/Artemis/commit/6b2b614f5aacf708689cb49a45d017b3d9c77a77

Author: Addressed with https://github.com/ls1intum/Artemis/commit/6b2b614f5aacf708689cb49a45d017b3d9c77a77

        vcs-credentials: artemis_localvc_credentials
        artemis-authentication-token-key: artemis_notification_plugin_token
        artemis-authentication-token-value: artemis_admin
        build-timeout: 30
    git:
        name: Artemis
        email: artemis@xcit.tum.de
   jenkins:

>        internal-urls:

Reviewer: ```suggestion
       # only required if Artemis and Jenkins cannot communicate on their public URLs
       # e.g., Jenkins is only available in a local container network
       internal-urls:
```

        `activated`,`lang_key`,`activation_key`,`reset_key`,`created_by`,`created_date`,`reset_date`,
        `last_modified_by`,`last_modified_date`,`image_url`,`last_notification_read`,`registration_number`)
        VALUES (1,"artemis_admin","HASHED_PASSWORD","artemis","administrator","artemis_admin@localhost",
        1,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
3. Give the user admin and user roles:

    .. code:: sql

        INSERT INTO `artemis`.`jhi_user_authority` (`user_id`, `authority_name`) VALUES (1,"ROLE_ADMIN");
        INSERT INTO `artemis`.`jhi_user_authority` (`user_id`, `authority_name`) VALUES (1,"ROLE_USER");

Starting the Artemis server should now succeed.

Jenkins
^^^^^^^

Automated Jenkins Server Setup
""""""""""""""""""""""""""""""

The following steps describe how to deploy a pre-configured version of the Jenkins server.
This is ideal as a quickstart for developers. For a more detailed setup, see
`Manual Jenkins Server Setup <#manual-jenkins-server-setup>`__.
In a production setup, you have to at least change the user credentials (in the file ``jenkins-casc-config-localvc.yml``) and push tokens.

> 1. You can now first build and deploy Jenkins, then you can also start the other services which weren't started yet:

Reviewer: ```suggestion
#. You can now first build and deploy Jenkins, then you can also start the other services which weren't started yet:
```
Same for other three items below to automatically get correct numbering. At the moment there are only steps 1, 2, 5, 6.

   Jenkins is then reachable under ``http://localhost:8082/`` and you can login using the credentials specified
   in ``jenkins-casc-config-localvc.yml`` (defaults to ``artemis_admin`` as both username and password).

2. The `application-local.yml` must be adapted with the values configured in ``jenkins-casc-config-localvc.yml``:

.. code:: yaml

    artemis:
        user-management:
            use-external: false
            internal-admin:
                username: artemis_admin
                password: artemis_admin
        version-control:
            url: http://172.17.0.1:8080 # `http://host.docker.internal:8080` for Windows
            user: root
            password: dummy # have to be set, but does not matter for LocalVC
            build-agent-git-username: jenkins
            build-agent-git-password: artemis_admin # choose some strong password and username (gives read access to all repositories)
        continuous-integration:
            user: artemis_admin
            password: artemis_admin
            url: http://localhost:8082
            vcs-credentials: artemis_localvc_credentials
            artemis-authentication-token-key: artemis_notification_plugin_token

>             artemis-authentication-token-value: artemis_admin

Reviewer: Should be indented once to be part of item 2 and make the automatic numbering work.

Author: Addressed with https://github.com/ls1intum/Artemis/commit/d8ca39e99d9de1ded6737ebe81423a73916b236b

5. Open the ``src/main/resources/config/application-jenkins.yml`` and change the following:
   Again, if you are using a development setup, the template in the beginning of this page already contains the
   correct values.

.. code:: yaml

    jenkins:
        internal-urls:
            ci-url: http://jenkins:8080

>             vcs-url: http://localhost:8080

Reviewer: Should be indented once to be part of the item and make the automatic numbering work.

Author: Addressed with https://github.com/ls1intum/Artemis/commit/d8ca39e99d9de1ded6737ebe81423a73916b236b

6. You're done. You can now run Artemis with the LocalVC/Jenkins environment.

Manual Jenkins Server Setup
"""""""""""""""""""""""""""

1. Pull the latest Jenkins LTS Docker image

   Run the following command to get the latest jenkins LTS docker image.

   .. code:: bash

       docker pull jenkins/jenkins:lts

2. Create a custom docker image

   In order to install and use Maven with Java in the Jenkins container,
   you have to first install maven, then download Java and finally
   configure Maven to use Java instead of the default version.
   You also need to install Swift and SwiftLint if you want to be able to
   create Swift programming exercises.

   To perform all these steps automatically, you can prepare a Docker
   image:

   Create a Dockerfile with the content found `here <docker/jenkins/Dockerfile>`.
   Copy it in a file named ``Dockerfile``, e.g. in
   the folder ``/opt/jenkins/`` using ``vim Dockerfile``.

   Now run the command ``docker build --no-cache -t jenkins-artemis .``

   This might take a while because Docker will download Java, but this

>    is only required once.

Reviewer: These steps can be removed: Jenkins no longer builds with Maven directly but always delegates the build to happen inside other containers.

> 
> 3. **If you run your own NGINX or if you install Jenkins on a local development computer, then skip the next steps (4-7)**
> 
> 4. Create a file increasing the maximum file size for the nginx proxy.

Reviewer: ```suggestion
Nginx proxy setup
#################

If you run your own NGINX or if you install Jenkins on a local development computer, you can skip this section.

#. Create a file increasing the maximum file size for the Nginx proxy.
```
Since the two steps above can be removed, this part only contains Nginx setup instructions now and could receive its own sub-heading.

       docker run --detach \
           --name nginx_proxy-letsencrypt \
           --volumes-from nginx_proxy \
           --volume /var/run/docker.sock:/var/run/docker.sock:ro \
           --env "DEFAULT_EMAIL=mail@yourdomain.tld" \
           jrcs/letsencrypt-nginx-proxy-companion

Start Jenkins
#############

8.  Run Jenkins by executing the following command (change the hostname
    and choose which port alternative you need)

    .. code:: bash

        docker run -itd --name jenkins \
            --restart always \
            -v jenkins_data:/var/jenkins_home \
            -v /var/run/docker.sock:/var/run/docker.sock \
            -v /usr/bin/docker:/usr/bin/docker:ro \
            -e VIRTUAL_HOST=your.jenkins.domain -e VIRTUAL_PORT=8080 \    # Alternative 1: If you are NOT using a separate NGINX instance
            -e LETSENCRYPT_HOST=your.jenkins.domain \                     # Only needed if Alternative 1 is used
            -p 8082:8080 \                                                # Alternative 2: If you ARE using a separate NGINX instance OR you ARE installing Jenkins on a local development computer
            -u root \
            jenkins/jenkins:lts

>     If you still need the old setup with Python & Maven installed locally, use ``jenkins-artemis`` instead of
>     ``jenkins/jenkins:lts``.
>     Also note that you can omit the ``-u root``, ``-v /var/run/docker.sock:/var/run/docker.sock`` and

Reviewer: ```suggestion
    Note that you can omit the ``-u root``, ``-v /var/run/docker.sock:/var/run/docker.sock`` and
```
No longer relevant, see comment above.
